datafusion/sql/src/values.rs

```diff
@@ -41,6 +44,11 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     .collect::<Result<Vec<_>>>()
             })
             .collect::<Result<Vec<_>>>()?;
-        LogicalPlanBuilder::values(values)?.build()
+
+        if schema.fields().is_empty() {
+            LogicalPlanBuilder::values(values, None)?.build()
+        } else {
+            LogicalPlanBuilder::values(values, Some(&schema))?.build()
```

Comment:
- is this in spirit similar to https://github.com/apache/datafusion/pull/12104 which didn't seem accepted?

- how do we know that VALUES being processed here should actually obey the schema of the "main table" of the query? For example:
   ```sql
  CREATE TABLE t AS SELECT a+1, 'foo' FROM (VALUES (3)) t(a)
  ```
  We have CREATE TABLE (which is the only case where `planner_context.table_schema()` exists), we have `VALUES`, but schema of the table, and schema of the values is not related.

Reply:
> CREATE TABLE t AS SELECT a+1, 'foo' FROM (VALUES (3)) t(a)

I think there is no schema in such query 🤔.

> create table t(a int) as values (1);

is the example that we have schema for values to infer the data type.


I don't quite follow what #12104 is doing, but I think it makes sense to cast the values if the table schema exist and coerce values with `type_union_resolution` if there is no given schema.

Reply:
> how do we know that VALUES being processed here should actually obey the schema of the "main table" of the query? 

My point here is that doing what we're doing here is legal only for certain query shapes involving CREATE TABLE + VALUES (eg example above), but is not applicable to other query shapes involving CREATE TABLE and VALUES (e.g. CREATE TABLE  +  SELECT .... FROM VALUES)

Where is the logic guaranteeing that table-inferred schema gets applied in the former case, but not in the latter?

Reply:
We can work on `CREATE TABLE + SELECT .... FROM VALUES` later on.